File: jumpstart/web/src/main/java/jumpstart/web/mixins/ZonePeriodicUpdater.js
// A class that periodically issues an AJAX request (spec.eventURL) to update a zone (spec.zoneElementId).
// Based on a solution by Taha Hafeez at
// http://tapestry.1045711.n5.nabble.com/T5-Complexity-for-simple-things-where-is-the-documentation-tt3214893.html

ZonePeriodicUpdater = Class.create( {

	initialize : function(spec) {
		this.zoneElement = $(spec.zoneElementId).getStorage();
		this.eventURL = spec.eventURL;
		this.frequencySecs = spec.frequencySecs;
		this.maxUpdates = spec.maxUpdates;

		// Use Prototype's PeriodicalExecuter to run updateZone() every 3 seconds, this.maxUpdates times.

		var updatesCount = 0;

		new PeriodicalExecuter(function(pe) {
			if (updatesCount++ < this.maxUpdates) {
				this.updateZone();
			}
			else {
				pe.stop();
			}
		}.bind(this), this.frequencySecs);

	},

	updateZone : function() {
		var zoneManager = this.zoneElement.zoneManager;
		if(!zoneManager){
			return;
		}

		zoneManager.updateFromURL(this.eventURL);
	}
	
} )

// Extend the Tapestry.Initializer with a static method that instantiates a ZonePeriodicUpdater.

Tapestry.Initializer.zonePeriodicUpdater = function(spec) {
    new ZonePeriodicUpdater(spec);
}

File: jumpstart/web/src/main/java/jumpstart/web/mixins/ZonePeriodicUpdater.java
/**
 * A simple mixin for attaching javascript to a Zone that periodically issues an AJAX request to update the Zone
 * Based on a solution by Taha Hafeez at
 * http://tapestry.1045711.n5.nabble.com/T5-Complexity-for-simple-things-where-is-the-documentation-tt3214893.html.
 */
package jumpstart.web.mixins;

import org.apache.tapestry5.BindingConstants;
import org.apache.tapestry5.ClientElement;
import org.apache.tapestry5.ComponentResources;
import org.apache.tapestry5.annotations.Import;
import org.apache.tapestry5.annotations.InjectContainer;
import org.apache.tapestry5.annotations.Parameter;
import org.apache.tapestry5.ioc.annotations.Inject;
import org.apache.tapestry5.json.JSONObject;
import org.apache.tapestry5.services.javascript.JavaScriptSupport;

// The @Import tells Tapestry to put a link to the file in the head of the page so that the browser will pull it in. 
@Import(library = "ZonePeriodicUpdater.js")
public class ZonePeriodicUpdater {
	
	// Parameters

	@Parameter(required = true, defaultPrefix = BindingConstants.LITERAL)
	private String event;

	@Parameter
	private Object[] context;

	@Parameter(defaultPrefix = BindingConstants.LITERAL, required = true)
	private int frequencySecs;

	@Parameter(defaultPrefix = BindingConstants.LITERAL, required = true)
	private int maxUpdates;

	// Useful bits and pieces

	@Inject
	private ComponentResources componentResources;

	@Inject
	private JavaScriptSupport javaScriptSupport;

	/**
	 * The element we attach ourselves to, which must be a Zone 
	 */
	@InjectContainer
	private ClientElement clientElement;

	// The code

	void afterRender() {

		// Tell the Tapestry.Initializer to do the initializing of a ZonePeriodicUpdater, which it will do when the DOM has
		// been fully loaded.

		JSONObject spec = new JSONObject();
		spec.put("zoneElementId", clientElement.getClientId());
		spec.put("eventURL", componentResources.createEventLink(event, context).toAbsoluteURI());
		spec.put("frequencySecs", frequencySecs);
		spec.put("maxUpdates", maxUpdates);
		javaScriptSupport.addInitializerCall("zonePeriodicUpdater", spec);
	}

}
